### ISSUE TITLE
IPSEC service restarts consistently after migrating from 4.1 -> 4.12

### ISSUE THREAD

**Issue by avaughn0909** [libreswan/libreswan#1609]
Hello again. So I had recently opened up issue #1574 about ipsec stuck in a crash loop when being sent unsolicited intermediate key exchange messages. As a part of the resolution to this issue, we received guidance to upgrade from 4.1 -> 4.12. We made a plan, we build a test-suite tool to create tunnels, and also worked with the remote tunnel that caused the first issue to test version 4.12 in our dev env. This all went as planned and we saw no issues except for these warning messges `warning IKEv2 liveness uses retransmit-timeout=, dpdtimeout= ignored`. When we went to do our update on 6 FEB 2024, we started having restart issues with our VPN instance. Looking at `ipsec auto --status`, we saw that connections would start to establish SAs with some connections while the rest would be stuck in `IKE_SA_INIT` after sending the initiator message and not responding to responder messages. After looking at the logs I noticed that there was this assert failure `ABORT: ASSERTION FAILED: *chosen_proposal == ((void *)0) (ikev2_process_sa_payload() +1143 /programs/pluto/ikev2_proposals.c)`. This seems like maybe we may have some sort of error within some of our configs but I don't know for sure. I am curious as to what happened and how we prevent it from happening again. Here is the CoreDump:
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/libexec/ipsec/pluto --leak-detective --config /etc/ipsec.conf --nofork'.
Program terminated with signal SIGABRT, Aborted.
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
[Current thread is 1 (Thread 0x7f8fb1dabb00 (LWP 60086))]
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f8fb314b859 in __GI_abort () at abort.c:79
#2  0x00005653a58cf6e8 in llog_passert (logger=0x5653a599e640 <global_logger>, where=where@entry=0x5653a5994690 <here>, fmt=fmt@entry=0x5653a592f82b "%s") at ./lib/libswan/passert.c:34
#3  0x00005653a58ac438 in ikev2_process_sa_payload (what=what@entry=0x5653a58ffbcf "IKE initiator (accepting)", sa_payload=0x5653a8186320, expect_ike=expect_ike@entry=true, expect_spi=expect_spi@entry=false,
    expect_accepted=expect_accepted@entry=true, opportunistic=<optimized out>, chosen_proposal=0x5653a7842a20, local_proposals=0x5653a6fcb9b8, logger=0x5653a7750a38) at ./programs/pluto/ikev2_proposals.c:1143
#4  0x00005653a581ef44 in process_v2_IKE_SA_INIT_response (ike=0x5653a7842418, unused_child=<optimized out>, md=0x5653a8186068) at ./programs/pluto/ikev2_ike_sa_init.c:1418
#5  0x00005653a58a405d in v2_dispatch (ike=0x5653a7842418, md=0x5653a8186068, svm=0x5653a59b0ba0 <v2_state_transition_table+448>) at ./programs/pluto/ikev2.c:2334
#6  0x00005653a581d5d4 in process_v2_IKE_SA_INIT (md=md@entry=0x5653a8186068) at ./programs/pluto/ikev2_ike_sa_init.c:484
#7  0x00005653a58a47b4 in ikev2_process_packet (md=md@entry=0x5653a8186068) at ./programs/pluto/ikev2.c:1847
#8  0x00005653a58b525d in process_md (md=0x5653a8186068) at ./programs/pluto/demux.c:185
#9  0x00005653a58b5a4a in process_iface_packet (fd=<optimized out>, ifp_arg=0x0, logger=<optimized out>) at ./programs/pluto/demux.c:293
#10 0x00005653a587a6ea in fd_read_listener_event_handler (fd=<optimized out>, events=<optimized out>, arg=<optimized out>) at ./programs/pluto/server.c:795
#11 0x00007f8fb33a5fbf in ?? () from /lib/x86_64-linux-gnu/libevent_core-2.1.so.7
#12 0x00007f8fb33a66ff in event_base_loop () from /lib/x86_64-linux-gnu/libevent_core-2.1.so.7
#13 0x00005653a587cab1 in run_server (conffile=0x5653a6cb10b8 "/etc/ipsec.conf", logger=0x7fff4398d520) at ./programs/pluto/server.c:1070
#14 0x00005653a581a848 in main (argc=<optimized out>, argv=<optimized out>) at ./programs/pluto/plutomain.c:1821
```
I am trying to find some useful logs but most of them just give me the assertion failure.

**Comment by letoams** [comment]
On Thu, 8 Feb 2024, Andrew Vaughn wrote:

> ABORT: ASSERTION FAILED: *chosen_proposal == ((void *)0)
> (ikev2_process_sa_payload() +1143 /programs/pluto/ikev2_proposals.c). This seems like maybe we may have some sort of error within
> some of our configs but I don't know for sure. I am curious as to what happened and how we prevent it from happening again. Here is
> the CoreDump:

This is a real bug, but we would need to see the debug log files to
trace down what happened.

It could be this: https://github.com/libreswan/libreswan/issues/505

But we also were lacking information to resolve that for sure. We
thought that one was fixed actually.

> I am trying to find some useful logs but most of them just give me the assertion failure.

It might require debugging. You can dynamically enable that without
restarting pluto by using: ipsec whack --debug-all

**Comment by cagney** [comment]
Can you try tweaking:
```
stf_status process_v2_IKE_SA_INIT_response(struct ike_sa *ike,
					   struct child_sa *unused_child UNUSED,
					   struct msg_digest *md)
```
just before:
```
		n = ikev2_process_sa_payload("IKE initiator (accepting)",
					     &sa_pd->pbs,
					     /*expect_ike*/ true,
					     /*expect_spi*/ false,
					     /*expect_accepted*/ true,
					     LIN(POLICY_OPPORTUNISTIC, c->policy),
					     &ike->sa.st_v2_accepted_proposal,
					     ike_proposals, ike->sa.st_logger);
```
adding something like:
```
if (!pexpect(ike->sa.st_v2_accepted_proposal == NULL)) {
  /* leak previous accepted proposal that shoudn't be there */
  ike->sa.st_v2_accepted_proposal = NULL;
}
```
and see if it triggers and stops the crash.
I the logs, can you look for IKE SAs getting back atypical responses such as invalid KE and COOKIE.

**Comment by cagney** [comment]
> It could be this: #505 

Unlikely.  That was for a rekeying Child SA.  Here the crash happens during IKE_SA_INIT.

**Comment by avaughn0909** [comment]
Thank you both for your responses, We currently had to downgrade back to 4.1 (Since we know it worked for almost all of our remote sites). We are looking to upgrade again in a maintenance window in a little over a week. This maybe an opportunity for us to turn this on. Do you happen to know what the latest version of LibreSwan would NOT be vulnerable to this bug? This will give our team something to think about and plan around.

**Comment by Joel-Robinson-Redox** [comment]
Here are the logs for one particular VPN tunnel that we're suspicious of. There are about 6 or 7 tunnels  that have errors in the logs, but this tunnel is the only one that has `Failed to find our PreShared Key` and  `STATE_V2_PARENT_I1 had internal error`.  This is a **heavily** redacted version of the logs that only shows this one customer connection plus the abort at end.
```
"Feb  7 03:08:15 ip-internal-iface pluto[57436]: ""customer1/2x15"": IKE SA proposals (connection add):"
"Feb  7 03:08:15 ip-internal-iface pluto[57436]: ""customer1/2x15"":   1:IKE=AES_CBC_256-HMAC_SHA2_256-HMAC_SHA2_256_128-ECP_384"
"Feb  7 03:08:15 ip-internal-iface pluto[57436]: ""customer1/2x15"": Child SA proposals (connection add):"
"Feb  7 03:08:15 ip-internal-iface pluto[57436]: ""customer1/2x15"":   1:ESP=AES_CBC_256-HMAC_SHA2_256_128-NONE-ENABLED+DISABLED"
"Feb  7 03:08:15 ip-internal-iface pluto[57436]: ""customer1/2x15"": added IKEv2 connection"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: initiating all conns with alias='customer1'"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: initiating IKEv2 connection"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x14"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x13"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x12"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x11"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x10"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x9"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x8"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x7"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x6"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x5"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x4"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x3"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x2"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x1"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x15"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x14"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x13"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x12"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x11"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x10"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x9"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x8"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x7"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x6"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x5"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x4"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x3"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x2"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/1x1"": queue Child SA; waiting on IKE SA ""customer1/2x15"" #119 negotiating with CUSTOMER_EXTERNAL_IP"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: sent IKE_SA_INIT request to CUSTOMER_EXTERNAL_IP:500"
"Feb  7 03:08:16 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: STATE_V2_PARENT_I1: retransmission; will wait 0.5 seconds for response"
"Feb  7 03:08:17 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: STATE_V2_PARENT_I1: retransmission; will wait 1 seconds for response"
"Feb  7 03:08:18 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: STATE_V2_PARENT_I1: retransmission; will wait 2 seconds for response"
"Feb  7 03:08:19 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: authentication failed: no PSK found for 'OUR_EXTERNAL_IP'"
"Feb  7 03:08:19 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: Failed to find our PreShared Key"
"Feb  7 03:08:19 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: state transition function for STATE_V2_PARENT_I1 had internal error"
"Feb  7 03:08:21 ip-internal-iface pluto[57436]: ABORT: ASSERTION FAILED: *chosen_proposal == ((void *)0) (ikev2_process_sa_payload() +1143 /programs/pluto/ikev2_proposals.c)"
```

**Comment by paulwouters** [comment]
I cannot reproduce this on git main:
```
initiating all conns with alias='road'
"road/0x2" #1: initiating IKEv2 connection
"road/0x1": queue Child SA; waiting on IKE SA "road/0x2" #1 negotiating with 192.1.2.23
"road/0x2" #1: sent IKE_SA_INIT request to 192.1.2.23:500
"road/0x2" #1: authentication failed: no PSK found for '@road'
"road/0x2" #1: Failed to find our PreShared Key
"road/0x2" #1: state transition function for STATE_V2_PARENT_I1 had internal error
"road/0x2" #1: STATE_V2_PARENT_I1: retransmission; will wait 0.5 seconds for response
```

No crash happens. This is using multiple subnets and a missing PSK

**Comment by cagney** [comment]
@Joel-Robinson-Redox tks:
```
"Feb  7 03:08:19 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: authentication failed: no PSK found for 'OUR_EXTERNAL_IP'"
"Feb  7 03:08:19 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: Failed to find our PreShared Key"
```
This indicates that IKE_AUTH processing has started (and IKE_SA_INIT response was finished, which is why `*chosen_proposal == ((void *)0)` is not NULL).  But then:
```
"Feb  7 03:08:19 ip-internal-iface pluto[57436]: ""customer1/2x15"" #119: state transition function for STATE_V2_PARENT_I1 had internal error"
```
Turning a missing key into an internal error, which turns `#119` into something like a zombie, is over kill.

I suspect, the peer has sent multiple IKE_SA_INIT responses due to the re-transmits, and one of those hit the zombie.

**Comment by cagney** [comment]
```
"westnet-eastnet-ipv4-psk-ikev2" #1: initiating IKEv2 connection to 192.1.2.23 using UDP
"westnet-eastnet-ipv4-psk-ikev2" #1: sent IKE_SA_INIT request to 192.1.2.23:UDP/500
"westnet-eastnet-ipv4-psk-ikev2" #1: authentication failed: no PSK found for '@west'
"westnet-eastnet-ipv4-psk-ikev2" #1: Failed to find our PreShared Key
"westnet-eastnet-ipv4-psk-ikev2" #1: state transition function for STATE_V2_PARENT_I1 had internal error
FATAL: ASSERTION FAILED: *chosen_proposal == ((void *)0) (ikev2_process_sa_payload() +1144 programs/pluto/ikev2_proposals.c)
```

**Comment by paulwouters** [comment]
As a workaround, you can add a line at the END of your ipsec.secrets file, eg:

`: PSK "somereallylongrandomunguessablestring"`

It has to appear after any include statement. This will prevent the crasher, as now we can always find a PSK that matches, and it will just be the wrong one and a regular authentication failed will happen.

We are working on a proper fix for the v4 and v5 code.

**Comment by Joel-Robinson-Redox** [comment]
Our normal workflow is basically just to append new entries to `ipsec.secrets`, so how import is it that this be the **last** entry. Depending on how matching works, it seems like it could hit this secret and ignore all subsequent entries.

**Comment by Joel-Robinson-Redox** [comment]
Sorry, I should have [read the ipsec.secrets man page](https://libreswan.org/man/ipsec.secrets.5.html). 
> To authenticate a connection between two hosts, the entry that most specifically matches the host and peer IDs is used. An entry with no index will match any host and peer. More specifically, an entry with one index will match a host and peer if the index matches the host's ID (the peer isn't considered). Still more specifically, an entry with multiple indices will match a host and peer if the host ID and peer ID each match one of the indices. It is acceptable for two entries to be the best match as long as they agree about the secret.

So it looks like it will only use the catch-all is nothing else matches.

**Comment by Joel-Robinson-Redox** [comment]
I'm trying to see if I can reproduce the error on my side and then also test the work-around.

Are these tests that would reproduce the errors on 4.12?
[ikev2-05-basic-psk-missing-initiator](https://github.com/libreswan/libreswan/tree/main/testing/pluto/ikev2-05-basic-psk-missing-initiator)
[ikev2-05-basic-psk-missing-responder](https://github.com/libreswan/libreswan/tree/main/testing/pluto/ikev2-05-basic-psk-missing-responder)

Found your [testing documentation](https://libreswan.org/wiki/Test_Suite), but any else related to testing that you could point me to would be appreciated.

**Comment by letoams** [comment]
On Mon, 26 Feb 2024, Joel Robinson wrote:

> I'm trying to see if I can reproduce the error on my side and then also test the work-around.
> 
> Are these tests that would reproduce the errors on 4.12?
> ikev2-05-basic-psk-missing-initiator
> ikev2-05-basic-psk-missing-responder
> 
> Found your testing documentation, but any else related to testing that you could point me to would be appreciated.

You would probably need to use an impair that delays the crypto
calculations to simulate the race condition of the retransmit.

Not sure if a new impair is needed or an existing one will work for
this. See include/impair.h and 'git grep IMPAIR'

**Comment by cagney** [comment]
> Are these tests that would reproduce the errors on 4.12? [ikev2-05-basic-psk-missing-initiator](https://github.com/libreswan/libreswan/tree/main/testing/pluto/ikev2-05-basic-psk-missing-initiator) [ikev2-05-basic-psk-missing-responder](https://github.com/libreswan/libreswan/tree/main/testing/pluto/ikev2-05-basic-psk-missing-responder)

The tests reproduced the bug in mainline, not 4.12.  They use an impair to send the delayed re-transmit see `04-east-retransmit.sh`:
```
ipsec whack --impair drip_outbound:1
```

> Found your [testing documentation](https://libreswan.org/wiki/Test_Suite), but any else related to testing that you could point me to would be appreciated.

I, manually, installed an old version of libreswan on `west` and then cut/paste the test commands.

**Comment by cagney** [comment]
fixed in v4.14 (technically v4.13)
